### PR TITLE
1) report_outcome() should log in background thread. 2) Bug fixes and cleanups

### PR DESCRIPTION
**Pull request by rajan-chari**
1) Move report outcome to background thread.  (feature)
2) Use object pool for buffer since this can be accessed from multiple threads.  (thread safety)
3) std::ostringstream reset does not work correctly. Replaced with data_buffer (bug fix)
4) use size_t for action id since it indexes an array.  (clean up compile)

### CODE REVIEW

**Comment by rajan-chari** [comment]
Removed basic_usage_cpp.out

File: reinforcement_learning/examples/rl_sim_cpp/rl_sim.cc [hunk 1]
```diff
@@ -157,7 +161,7 @@ rl_sim::rl_sim(boost::program_options::variables_map vm) :_options(std::move(vm)
 std::string get_dist_str(const reinforcement_learning::ranking_response& response) {
   std::string ret;
   ret += "(";
-  for (auto& ap_pair : response) {
+  for (auto ap_pair : response) {
```
Comment (ataymano) [inline reinforcement_learning/examples/rl_sim_cpp/rl_sim.cc]: const auto& ?

Comment (rajan-chari) [reply reinforcement_learning/examples/rl_sim_cpp/rl_sim.cc]: Makes sense.

File: reinforcement_learning/rlclientlib/utility/data_buffer.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+#include <string>
+#include <vector>
+
+namespace reinforcement_learning { namespace utility {
+  class data_buffer {
+  public:
+
+    void reset();
+    std::string str();
```
Comment (ataymano) [inline reinforcement_learning/rlclientlib/utility/data_buffer.h]: const?

Comment (rajan-chari) [reply reinforcement_learning/rlclientlib/utility/data_buffer.h]: Makes sense.

**Pull request merged**